cdap-app-fabric/src/main/java/co/cask/cdap/reporting/ProgramHeartbeatDataset.java

```diff
@@ -0,0 +1,205 @@
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Heartbeat Store that writes heart beat messages and program status messages
+ * to program heartbeat table. This is used for efficiently
+ * scanning and returning results for dashboard status queries
+ *
+ * Rowkey for the table is organized in the following format
+ * [namespace][row-key-separator][timestamp-in-seconds][program-run-Id]
+ *
+ */
+public class ProgramHeartbeatDataset extends AbstractDataset {
+  private static final Gson GSON = TriggeringScheduleInfoAdapter.addTypeAdapters(new GsonBuilder()).create();
+  private static final byte[] COLUMN_NAME = Bytes.toBytes("status");
+  private static final DatasetId PROGRAM_HEARTBEAT_INSTANCE_ID =
+    NamespaceId.SYSTEM.dataset(Constants.ProgramHeartbeat.TABLE);
+  private static final byte[] ROWKEY_SEPARATOR = Bytes.toBytes(":");
+  private final Table table;
+
+  public ProgramHeartbeatDataset(Table table) {
+    super("ignored", table);
+    this.table = table;
+  }
+
+  /**
+   * Static method for creating an instance of {@link ProgramHeartbeatDataset}.
+   */
+  public static ProgramHeartbeatDataset getOrCreate(DatasetContext datasetContext,
+                                                    DatasetFramework datasetFramework, CConfiguration cConfiguration) {
+    try {
+      TableProperties.Builder props = TableProperties.builder();
+      long ttl = cConfiguration.getLong(Constants.ProgramHeartbeat.HEARTBEAT_TABLE_TTL_SECONDS);
+      props.setTTL(ttl);
+      Table table = DatasetsUtil.getOrCreateDataset(datasetContext, datasetFramework, PROGRAM_HEARTBEAT_INSTANCE_ID,
+                                                    Table.class.getName(), props.build());
+      return new ProgramHeartbeatDataset(table);
+    } catch (DatasetManagementException | IOException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  /**
+   * Write {@link RunRecordMeta} to heart beat dataset as value,
+   * with rowKey in format [namespace][row-key-separator][timestamp-in-seconds][program-run-Id]
+   *
+   * @param runRecordMeta row value to write
+   * @param timestampInSeconds used for creating rowKey
+   */
+  public void writeRunRecordMeta(RunRecordMeta runRecordMeta, long timestampInSeconds) {
+    byte[] rowKey = createRowKey(timestampInSeconds, runRecordMeta.getProgramRunId());
+    table.put(rowKey, COLUMN_NAME, Bytes.toBytes(GSON.toJson(runRecordMeta)));
+  }
+
+  private byte[] createRowKey(long timestampInSeconds, ProgramRunId programRunId) {
+    MDSKey.Builder builder = new MDSKey.Builder();
+    // add namespace at the beginning
+    builder.add(programRunId.getNamespace());
+    // add row key separator
+    // we add a rowkey separator between namespace and timestamp as namespace has variable length and there could also
+    // be namespace which might have timestamp as its name,
+    // having a separator helps to provide accurate result in those scenarios
+    builder.add(ROWKEY_SEPARATOR);
```

Comment:
you don't need separators with MDSKey. MDSKey writes the length of the part, and then the part itself.